stdlib/public/SDK/Foundation/IndexPath.swift

```diff
@@ -124,78 +622,115 @@ public struct IndexPath : ReferenceConvertible, Equatable, Hashable, MutableColl
     public var endIndex: Index {
         return _indexes.endIndex
     }
-
+    
     public func index(before i: Index) -> Index {
         return _indexes.index(before: i)
     }
     
     public func index(after i: Index) -> Index {
         return _indexes.index(after: i)
     }
-
+    
     /// Sorting an array of `IndexPath` using this comparison results in an array representing nodes in depth-first traversal order.
     public func compare(_ other: IndexPath) -> ComparisonResult  {
-        // This is temporary
-        let me = self.makeReference()
-        let other = other.makeReference()
-        return me.compare(other as IndexPath)
+        let thisLength = count
+        let otherLength = other.count
+        let length = Swift.min(thisLength, otherLength)
+        for idx in 0..<length {
+            let otherValue = other[idx]
+            let value = self[idx]
+            if value < otherValue {
+                return .orderedAscending
+            } else if value > otherValue {
+                return .orderedDescending
+            }
+        }
+        if thisLength > otherLength {
+            return .orderedDescending
+        } else if thisLength < otherLength {
+            return .orderedAscending
+        }
+        return .orderedSame
     }
-
+    
     public var hashValue: Int {
-        // This is temporary
-        let me = self.makeReference()
-        return me.hash
+        func hashIndexes(first: Int, last: Int, count: Int) -> Int {
+            let totalBits = MemoryLayout<Int>.size * 8
+            let lengthBits = 8
+            let firstIndexBits = (totalBits - lengthBits) / 2
+            return count + (first << lengthBits) + (last << (lengthBits + firstIndexBits))
```

Comment:
Shouldn't this have been:

```swift
return count &+ (first << lengthBits) &+ (last << (lengthBits + firstIndexBits))
```

i.e. if the first/last value are large, you can overflow.